File: python/libid_vibe.py
"""
Randomized linear-algebra utilities.

This module implements several randomized routines that approximate the
column space, QR factorization, singular-value decomposition (SVD), and
interpolative decomposition (ID) of a matrix A.

Public functions
----------------
orth_sketch        - Build an orthonormal basis for the column space.
rrqr_randomized    - Rank-revealing QR using a randomized basis.
rrsvd_randomized   - Truncated SVD using a randomized basis.
rrid_randomized    - Interpolative decomposition using randomized QR.

Author: Your Name
SPDX-License-Identifier: TBD
"""

import numpy as np
from scipy import linalg
from numpy.linalg import norm


def _power_iteration(A, X, flag_power=0):
    """
    Apply `flag_power` steps of power iteration to improve the sketch.

    Parameters
    ----------
    A : ndarray
        Input matrix.
    X : ndarray
        Random sketch matrix.
    flag_power : int, optional
        Number of power-iteration steps. Default is 0 (no iteration).

    Returns
    -------
    ndarray
        Updated sketch matrix.
    """
    for _ in range(flag_power):
        X = A.conj().T @ (A @ X)
        X, _, _ = linalg.qr(x, mode="economic", pivoting=True)
    return X


def orth_sketch(A, rtol, block_size=42, flag_power=0):
    """
    Build an orthonormal basis for the column space of A.

    The algorithm draws a random matrix, optionally improves it with
    power iteration, and then performs a QR factorization with column
    pivoting.  The process repeats with a larger sketch until the
    smallest diagonal element of R, relative to the column norms,
    falls below `rtol`.

    Parameters
    ----------
    A : ndarray
        Input matrix of shape (m, n).
    rtol : float
        Relative tolerance that controls the stopping criterion.
    block_size : int, optional
        Initial number of random vectors. Default is 42.
    flag_power : int, optional
        Number of power-iteration steps. Default is 0.

    Returns
    -------
    int
        Effective rank (number of basis vectors).
    ndarray
        Orthonormal basis matrix Q of shape (m, rank). If the rank
        equals min(m, n) an empty array with shape (0, 0) is returned.
    """
    m, n = A.shape

    if block_size >= min(m, n):
        return min(m, n), np.empty_like(A, shape=(0, 0))

    while True:
        # Random matrix with entries in [-1, 1].
        X = (2 * np.random.uniform(size=(n, block_size)) - 1).astype(A.dtype)
        X = _power_iteration(A, X, flag_power)
        y = A @ X

        # QR with column pivoting.
        Q, R, _ = linalg.qr(y, mode="economic", pivoting=True)
        r_diag = R.diagonal()
        # Smallest diagonal entry relative to the largest column norm.
        residual = max(abs(r_diag[-1:])) / max(norm(y, axis=0))

        if residual <= rtol:
            return block_size, Q

        # If not good enough, increase the sketch size.
        block_size = min(block_size * 4, min(m, n))

        if block_size >= min(m, n):
            return min(m, n), np.empty_like(A, shape=(0, 0))


def rrqr_randomized(A, rtol, block_size=42, flag_power=0):
    """
    Rank-revealing QR factorization using a randomized sketch.

    Parameters
    ----------
    A : ndarray
        Input matrix of shape (m, n).
    rtol : float
        Relative tolerance for determining numerical rank.
    block_size : int, optional
        Initial sketch size. Default is 42.
    flag_power : int, optional
        Number of power-iteration steps. Default is 0.

    Returns
    -------
    Q : ndarray
        Orthonormal matrix of shape (m, k) where k is the estimated rank.
    R : ndarray
        Upper-triangular factor of shape (k, n).
    p : ndarray
        Pivot indices such that A[:, p] \approx Q @ R.
    """
    m, n = A.shape
    k, q = orth_sketch(A, rtol, block_size, flag_power)

    if k >= min(m, n):
        Q, R, p = linalg.qr(A, mode="economic", pivoting=True)
        k = np.sum(norm(R, axis=1) >= rtol * norm(A))
        return Q[:, :k], R[:k, :], p

    # Project A onto the sketch subspace.
    Aproj = q.conj().T @ A
    Qproj, R, p = linalg.qr(Aproj, mode="economic", pivoting=True)
    Q = q @ Qproj
    k = np.sum(norm(R, axis=1) >= rtol * norm(Aproj))
    return Q[:, :k], R[:k, :], p


def rrsvd_randomized(A, rtol, block_size=42, flag_power=0):
    """
    Truncated singular-value decomposition using a randomized sketch.

    Parameters
    ----------
    A : ndarray
        Input matrix of shape (m, n).
    rtol : float
        Relative tolerance for determining the retained singular values.
    block_size : int, optional
        Initial sketch size. Default is 42.
    flag_power : int, optional
        Number of power-iteration steps. Default is 0.

    Returns
    -------
    U : ndarray
        Left singular vectors of shape (m, k).
    s : ndarray
        Singular values (length k).
    Vt : ndarray
        Right singular vectors transposed, shape (k, n).
    """
    m, n = A.shape
    k, q = orth_sketch(A, rtol, block_size, flag_power)

    if k >= min(m, n):
        U, s, Vt = linalg.svd(A, full_matrices=False)
        k = np.sum(np.abs(s) >= rtol * norm(A))
        return U[:, :k], s[:k], Vt[:k, :]

    Aproj = q.conj().T @ A
    Uproj, s, Vt = linalg.svd(Aproj, full_matrices=False)
    U = q @ Uproj
    k = np.sum(np.abs(s) >= rtol * norm(Aproj))
    return U[:, :k], s[:k], Vt[:k, :]


def rrid_randomized(A, rtol, block_size=42, flag_power=0):
    """
    Interpolative decomposition using a randomized QR factorization.

    Parameters
    ----------
    A : ndarray
        Input matrix of shape (m, n).
    rtol : float
        Relative tolerance for rank determination.
    block_size : int, optional
        Initial sketch size. Default is 42.
    flag_power : int, optional
        Number of power-iteration steps. Default is 0.

    Returns
    -------
    k : int
        Estimated rank.
    p : ndarray
        Pivot indices such that the first `k` columns of A[:, p] form a
        basis for the column space.
    T : ndarray
        Interpolation matrix satisfying
        A[:, p[k:]] \approx A[:, p[:k]] @ T.
    """
    Q, R, p = rrqr_randomized(A, rtol, block_size, flag_power)
    k = R.shape[0]
    # Solve R11 * T = R12 for T.
    T = linalg.solve(np.triu(R[:k, :k]), R[:k, k:])
    return k, p, T


def _hilb(n, m):
    """
    Create an n-by-m Hilbert matrix.

    Parameters
    ----------
    n : int
        Number of rows.
    m : int
        Number of columns.

    Returns
    -------
    ndarray
        Hilbert matrix of shape (n, m).
    """
    # Build first column and last row for the Hankel matrix.
    c = np.zeros(n)
    r = np.zeros(m)

    for i in range(n):
        c[i] = 1.0 / (i + 1)          # 0-based indexing

    for i in range(m):
        r[i] = 1.0 / (i + n)          # 0-based indexing

    return linalg.hankel(c, r)


if __name__ == "__main__":
    # Simple sanity checks for the public API.
    np.random.seed(0)

    # Small test matrix.
    m, n = 4000, 2000
    A = _hilb(m, n)

    # --------------------------------------------------------------
    # Test orth_sketch
    # --------------------------------------------------------------
    k_range, Q_range = orth_sketch(A, rtol=1e-12)
    orth_err = np.linalg.norm(Q_range.conj().T @ Q_range - np.eye(k_range))
    print(f"orth_sketch: k={k_range}, basis shape={Q_range.shape}")
    print(f"orth_sketch: orthonormality error={orth_err:e}")

    # --------------------------------------------------------------
    # Test rrqr_randomized
    # --------------------------------------------------------------
    Q_rrqr, R_rrqr, piv = rrqr_randomized(A, rtol=1e-12)
    A_perm = A[:, piv]
    recon_err = np.linalg.norm(Q_rrqr @ R_rrqr - A_perm) / np.linalg.norm(A_perm)
    print(f"rrqr_randomized: reconstruction relative error={recon_err:e}")

    # --------------------------------------------------------------
    # Test rrsvd_randomized
    # --------------------------------------------------------------
    U_rrsvd, s_rrsvd, Vt_rrsvd = rrsvd_randomized(A, rtol=1e-12)
    A_svd = U_rrsvd @ np.diag(s_rrsvd) @ Vt_rrsvd
    svd_err = np.linalg.norm(A_svd - A) / np.linalg.norm(A)
    print(f"rrsvd_randomized: reconstruction relative error={svd_err:e}")

    # --------------------------------------------------------------
    # Test rrid_randomized
    # --------------------------------------------------------------
    k_id, piv_id, T_id = rrid_randomized(A, rtol=1e-12)
    A_id_approx = A[:, piv_id[:k_id]] @ T_id
    id_err = np.linalg.norm(A[:, piv_id[k_id:]] - A_id_approx) / np.linalg.norm(A)
    print(f"rrid_randomized: interpolation relative error={id_err:e}")

File: python/libid_vibe.m
% Description
% ----------
% This class implements several randomized linear‑algebra routines that
% approximate the rank, QR factorization, singular‑value decomposition
% (SVD), and interpolative decomposition (ID) of a matrix A.
%
% User‑callable methods
% ---------------------
%   range_randomized   - Build an orthonormal basis for the column space.
%   rrqr_randomized    - Rank‑revealing QR using a randomized basis.
%   rrsvd_randomized   - Truncated SVD using a randomized basis.
%   rrid_randomized    - Interpolative decomposition using randomized QR.
%   image_randomized   - Basis for the row space via transpose.
%
% Author: Your Name
% SPDX-License-Identifier: TBD

classdef RandomizedLinearAlgebra
    % RandomizedLinearAlgebra
    %
    % Collection of static methods that perform randomized matrix factorizations.
    %
    % See also: range_randomized, rrqr_randomized, rrsvd_randomized,
    %           rrid_randomized, image_randomized

    methods (Static)

        function [k, Q] = range_randomized(A, rtol, block_size, flag_power)
        % RANGE_RANDOMIZED Compute an orthonormal basis for the column space of A using random sampling.
        %
        % Description
        % -----------
        % Build an orthonormal basis for the column space of A using random
        % sampling and optional power iterations.  The routine stops when
        % the relative residual falls below ``rtol``.
        %
        % Parameters
        % ----------
        % A          : double matrix
        %   Input matrix.
        % rtol       : double
        %   Relative tolerance that determines when to stop sampling.
        % block_size : int, optional (default = 42)
        %   Initial number of random vectors.
        % flag_power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % k          : int
        %   Number of basis vectors found (may equal min(m,n)).
        % Q          : double matrix
        %   Orthonormal basis matrix with size (m,k).  If k == 0 the array is empty.
        %
        % Notes
        % -----
        % 1. If the initial block already covers the whole space, the function
        %    returns early.
        % 2. The random matrix X has entries in [-1,1].
        % 3. Power iteration is performed by the private method powerIteration.
        %
        % Example
        % -------
        %   A = randn(100,50);
        %   [k,Q] = RandomizedLinearAlgebra.range_randomized(A,1e-6);
        %
        %   -------------------------------------------------
        %   Code flow
        %   -------------------------------------------------
        %   1. Determine matrix dimensions.
        %   2. Check early‑exit condition.
        %   3. Loop:
        %        a) Generate random test matrix X.
        %        b) Apply power iteration (if flag_power > 0).
        %        c) Form Y = A*X and compute its QR factorization.
        %        d) Estimate residual and compare with rtol.
        %        e) Increase block size if needed.
        %   4. Return block size and orthonormal basis Q.
        %   -------------------------------------------------
        %
        % See also: rrqr_randomized, rrsvd_randomized, rrid_randomized

            if nargin < 4, flag_power = 0; end
            if nargin < 3, block_size = 42; end

            [m, n] = size(A);

            % If the initial block already covers the whole space, return early.
            if block_size >= min(m,n)
                k = min(m,n);
                Q = [];
                return
            end

            while true
                % Random matrix with entries in [-1,1]
                X = 2*rand(n, block_size) - 1;
                X = RandomizedLinearAlgebra.powerIteration(A, X, flag_power);

                Y = A * X;
                [Qtmp,R,~] = qr(Y,0);   % economy QR

                % Use the last diagonal entry of R as a proxy for the residual.
                rdiag = diag(R);
                residual = max(abs(rdiag(end))) / max(vecnorm(Y,2,1));

                if residual <= rtol
                    k = block_size;
                    Q = Qtmp;
                    return
                end

                % If residual is too large, increase the block size.
                block_size = min(block_size*4, min(m,n));

                if block_size >= min(m,n)
                    k = min(m,n);
                    Q = [];
                    return
                end
            end
        end


        function [Qk, Rk, p] = rrqr_randomized(A, rtol, block_size, flag_power)
        % RRQR_RANDOMIZED Rank‑revealing QR factorization using a randomized basis.
        %
        % Description
        % -----------
        % Compute a rank‑revealing QR factorization of A by first building a
        % randomized orthonormal basis for the column space.  If the matrix
        % is effectively full rank, a deterministic QR is performed.
        %
        % Parameters
        % ----------
        % A          : double matrix
        %   Input matrix.
        % rtol       : double
        %   Relative tolerance for rank determination.
        % block_size : int, optional (default = 42)
        %   Initial number of random vectors.
        % flag_power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % Qk         : double matrix
        %   Leading k columns of the orthogonal factor.
        % Rk         : double matrix
        %   Leading k rows of the upper‑triangular factor.
        % p          : int vector
        %   Pivot permutation vector.
        %
        % Notes
        % -----
        % 1. The rank k is chosen as the number of rows of R whose 2‑norm
        %    exceeds ``rtol * ||A||`` (or ``||A_proj||`` for the projected case).
        % 2. The private method ``powerIteration`` is used internally.
        %
        % Example
        % -------
        %   A = randn(200,80);
        %   [Q,R,p] = RandomizedLinearAlgebra.rrqr_randomized(A,1e-8);
        %
        %   -------------------------------------------------
        %   Code flow
        %   -------------------------------------------------
        %   1. Call range_randomized to obtain basis Q_basis.
        %   2. If full rank, compute deterministic QR of A.
        %   3. Otherwise project A onto the basis and QR the small matrix.
        %   4. Determine numerical rank k from R.
        %   5. Return truncated factors and pivot vector.
        %   -------------------------------------------------
        %
        % See also: range_randomized, rrsvd_randomized, rrid_randomized

            if nargin < 4, flag_power = 0; end
            if nargin < 3, block_size = 42; end

            [m, n] = size(A);
            [k, Q_basis] = RandomizedLinearAlgebra.range_randomized(A, rtol, block_size, flag_power);

            if k >= min(m,n)
                [Q,R,E] = qr(A,0);
                p = E;
                k = sum(vecnorm(R,2,2) >= rtol*norm(A,'fro'));
                Qk = Q(:,1:k);
                Rk = R(1:k,:);
                return
            end

            % Project A onto the basis and factor the small matrix.
            A_proj = Q_basis' * A;
            [Q_proj,R,E] = qr(A_proj,0);
            p = E;
            Qk = Q_basis * Q_proj;
            k = sum(vecnorm(R,2,2) >= rtol*norm(A_proj,'fro'));
            Qk = Qk(:,1:k);
            Rk = R(1:k,:);
        end


        function [Uk, sk, Vk] = rrsvd_randomized(A, rtol, block_size, flag_power)
        % RRSVD_RANDOMIZED Truncated singular‑value decomposition using a randomized basis.
        %
        % Description
        % -----------
        % Compute a truncated SVD of A by first constructing a randomized
        % orthonormal basis for the column space.  If A is effectively full
        % rank, the full deterministic SVD is performed.
        %
        % Parameters
        % ----------
        % A          : double matrix
        %   Input matrix.
        % rtol       : double
        %   Relative tolerance for truncation.
        % block_size : int, optional (default = 42)
        %   Initial number of random vectors.
        % flag_power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % Uk         : double matrix
        %   Leading k left singular vectors.
        % sk         : double vector
        %   Leading k singular values.
        % Vk         : double matrix
        %   Leading k right singular vectors (rows of V^H).
        %
        % Notes
        % -----
        % 1. The rank k is the number of singular values greater than
        %    ``rtol * ||A||`` (or ``||A_proj||`` for the projected case).
        %
        % Example
        % -------
        %   A = randn(150,120);
        %   [U,S,V] = RandomizedLinearAlgebra.rrsvd_randomized(A,1e-7);
        %
        %   -------------------------------------------------
        %   Code flow
        %   -------------------------------------------------
        %   1. Obtain basis Q_basis via range_randomized.
        %   2. If full rank, call deterministic svd.
        %   3. Otherwise form A_proj = Q_basis' * A.
        %   4. Compute svd of the small matrix.
        %   5. Lift left singular vectors back: U = Q_basis * U_proj.
        %   6. Truncate to k based on rtol.
        %   -------------------------------------------------
        %
        % See also: range_randomized, rrqr_randomized, rrid_randomized

            if nargin < 4, flag_power = 0; end
            if nargin < 3, block_size = 42; end

            [m, n] = size(A);
            [k, Q_basis] = RandomizedLinearAlgebra.range_randomized(A, rtol, block_size, flag_power);

            if k >= min(m,n)
                [U,S,V] = svd(A, 'econ');
                k = sum(abs(diag(S)) >= rtol*norm(A,'fro'));
                Uk = U(:,1:k);
                sk = diag(S(1:k,1:k));
                Vk = V(:,1:k)';
                return
            end

            A_proj = Q_basis' * A;
            [U_proj,S_proj,V_proj] = svd(A_proj, 'econ');
            k = sum(abs(diag(S_proj)) >= rtol*norm(A_proj,'fro'));

            Uk = Q_basis * U_proj(:,1:k);
            sk = diag(S_proj(1:k,1:k));
            Vk = V_proj(:,1:k)';
        end


        function [k, p, proj] = rrid_randomized(A, rtol, block_size, flag_power)
        % RRID_RANDOMIZED Interpolative decomposition using a randomized QR factorization.
        %
        % Description
        % -----------
        % Form an interpolative decomposition (ID) of A by first computing a
        % randomized rank‑revealing QR and then solving a triangular system to
        % obtain the interpolation matrix.
        %
        % Parameters
        % ----------
        % A          : double matrix
        %   Input matrix.
        % rtol       : double
        %   Relative tolerance for rank determination.
        % block_size : int, optional (default = 42)
        %   Initial number of random vectors.
        % flag_power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % k          : int
        %   Numerical rank (size of R11).
        % p          : int vector
        %   Pivot permutation vector.
        % proj       : double matrix
        %   Interpolation matrix such that A(:,p) ≈ A(:,p(1:k))*proj.
        %
        % Notes
        % -----
        % 1. The method uses the private ``powerIteration`` routine indirectly
        %    through ``rrqr_randomized``.
        %
        % Example
        % -------
        %   A = randn(80,200);
        %   [k,p,proj] = RandomizedLinearAlgebra.rrid_randomized(A,1e-6);
        %
        %   -------------------------------------------------
        %   Code flow
        %   -------------------------------------------------
        %   1. Call rrqr_randomized to obtain Q, R, and pivot vector p.
        %   2. Extract R11 (upper‑triangular leading block) and R12.
        %   3. Solve R11 * X = R12 for the interpolation matrix X.
        %   4. Return rank k, pivot vector, and X.
        %   -------------------------------------------------
        %
        % See also: rrqr_randomized, rrsvd_randomized, range_randomized

            if nargin < 4, flag_power = 0; end
            if nargin < 3, block_size = 42; end

            [Q,R,p] = RandomizedLinearAlgebra.rrqr_randomized(A, rtol, block_size, flag_power);
            k = size(R,1);

            % Solve R11 * X = R12 for X, where R = [R11 R12].
            R11 = triu(R(1:k,1:k));
            R12 = R(1:k,k+1:end);
            proj = R11 \ R12;
        end


        function [k, Q] = image_randomized(A, rtol, block_size, flag_power)
        % IMAGE_RANDOMIZED Compute a basis for the row space of A by applying range_randomized to the transpose.
        %
        % Description
        % -----------
        % A thin wrapper that calls ``range_randomized`` on ``A.'`` to obtain
        % a basis for the row space.
        %
        % Parameters
        % ----------
        % A          : double matrix
        %   Input matrix.
        % rtol       : double
        %   Relative tolerance for stopping criterion.
        % block_size : int, optional (default = 42)
        %   Initial number of random vectors.
        % flag_power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % k          : int
        %   Number of basis vectors found.
        % Q          : double matrix
        %   Orthonormal basis for the row space (size n‑by‑k).
        %
        % Example
        % -------
        %   A = randn(60,120);
        %   [k,Q] = RandomizedLinearAlgebra.image_randomized(A,1e-5);
        %
        % See also: range_randomized

            if nargin < 4, flag_power = 0; end
            if nargin < 3, block_size = 42; end

            [k, Q] = RandomizedLinearAlgebra.range_randomized(A.', rtol, block_size, flag_power);
        end

    end

    methods (Static, Access = private)

        function X = powerIteration(A, X, power)
        % POWERITERATION Apply power iteration to improve the quality of the sampling matrix.
        %
        % Description
        % -----------
        % Multiply the test matrix X by A and A' repeatedly to amplify the
        % dominant singular directions.  After each iteration a QR factorization
        % re‑orthogonalizes X.
        %
        % Parameters
        % ----------
        % A     : double matrix
        %   Input matrix.
        % X     : double matrix
        %   Random test matrix.
        % power : int, optional (default = 0)
        %   Number of power‑iteration steps.
        %
        % Returns
        % -------
        % X     : double matrix
        %   Updated test matrix after power iteration.
        %
        % Notes
        % -----
        % This routine is used internally by ``range_randomized``.
        %
        % Example
        % -------
        %   X = randn(100,20);
        %   X = RandomizedLinearAlgebra.powerIteration(A, X, 2);
        %
        % See also: range_randomized

            if nargin < 3, power = 0; end

            for ii = 1:power
                X = A.' * (A * X);
                [X,~,~] = qr(X,0);
            end
        end

    end
end

